packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,391 @@
+    })
+    ;(newState as any).contractCache = new Map(this.contractCache)
+    ;(newState as any).storageCache = new Map(this.storageCache)
+    ;(newState as any)._cache = this._cache
+    return newState
+  }
+
+  /**
+   * Sets the new block tag used when querying the provider and clears the
+   * internal cache.
+   * @param blockTag - the new block tag to use when querying the provider
+   */
+  setBlockTag(blockTag: BlockTagType): void {
+    if (typeof blockTag === 'bigint') {
+      this.blockTag = bigIntToHex(blockTag)
+    } else {
+      this.blockTag = blockTag
+    }
+    this.clearCache()
+  }
+
+  /**
+   * Clears the internal cache so all accounts, contract code, and storage slots will
+   * initially be retrieved from the provider
+   */
+  clearCache(): void {
+    this.contractCache.clear()
+    this.storageCache.clear()
+    this._cache.clear()
+  }
+
+  /**
+   * Gets the code corresponding to the provided `address`.
+   * @param address - Address to get the `code` for
+   * @returns {Promise<Buffer>} - Resolves with the code corresponding to the provided address.
+   * Returns an empty `Buffer` if the account has no associated code.
+   */
+  async getContractCode(address: Address): Promise<Buffer> {
+    let codeBuffer = this.contractCache.get(address.toString())
+    if (codeBuffer !== undefined) return codeBuffer
+    const code = await this.provider.getCode(address.toString(), this.blockTag)
+    codeBuffer = toBuffer(code)
+    this.contractCache.set(address.toString(), codeBuffer)
+    return codeBuffer
+  }
+
+  /**
+   * Adds `value` to the state trie as code, and sets `codeHash` on the account
+   * corresponding to `address` to reference this.
+   * @param address - Address of the `account` to add the `code` for
+   * @param value - The value of the `code`
+   */
+  async putContractCode(address: Address, value: Buffer): Promise<void> {
+    // Store contract code in the cache
+    this.contractCache.set(address.toString(), value)
+  }
+
+  /**
+   * Gets the storage value associated with the provided `address` and `key`. This method returns
+   * the shortest representation of the stored value.
+   * @param address - Address of the account to get the storage for
+   * @param key - Key in the account's storage to get the value for. Must be 32 bytes long.
+   * @returns {Buffer} - The storage value for the account
+   * corresponding to the provided address at the provided key.
+   * If this does not exist an empty `Buffer` is returned.
+   */
+  async getContractStorage(address: Address, key: Buffer): Promise<Buffer> {
+    // Check storage slot in cache
+    const accountStorage: Map<string, Buffer> | undefined = this.storageCache.get(
+      address.toString()
+    )
+    let storage: Buffer | string | undefined
+    if (accountStorage !== undefined) {
+      storage = accountStorage.get(key.toString('hex'))
+      if (storage !== undefined) {
+        return storage
+      }
+    }
```

Comment:
I think we'd encounter issues in some scenarios, for example the following (maybe that's an intentional limitation of the current ethersStateManager):

- `this.blockTag` = `"latest"`
- we call `getContractStorage()` with an arbitrary address/key pair (it's therefore retrieved at the state of the latest block)
- some time passes, and that address/key pair storage is updated 
- we call `getContractStorage()` again with the same address/key pair, the (outdated) cache is retrieved and returned 

What if we bypassed the cache is one of the "dynamic" values (pending, earliest, latest)?

Reply:
I've updated the readme section related to this state manager to hopefully clarify these points.

Reply:
I think the point is valid, between consecutive calls of fetches in race conditions for the same run, dynamic tag will bring inconsistent data, the easy way to handle is when some one `inits`/`setTag` the "run" starts and the etherstatemanager should get the block corresponding to that and run it (I am not sure how pending can be supported unless the node is not syncing more blocks, and also not sure about getProofs for pending)

Reply:
Resolved by removing the option to pass in dynamic block tags.

Reply:
I don't think people should be allowed to use `latest` or `pending` since those are dynamic; if inbetween calls to ethers the block changes then you get incorrect results.